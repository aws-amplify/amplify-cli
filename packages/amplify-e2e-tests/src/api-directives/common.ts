import path from 'path';
import fs from 'fs-extra';
import _ from 'lodash';
import gql from 'graphql-tag';
import { addApi, amplifyPush, updateAuthAddUserGroups } from 'amplify-e2e-core';

import {
  setupUser,
  getUserPoolId,
  getApiKey,
  configureAmplify,
  signInUser,
  getConfiguredAppsyncClientAPIKeyAuth,
  getConfiguredAppsyncClientCognitoAuth,
} from './authHelper';

const GROUPNAME = 'Admin';
const USERNAME = 'user1';
const PASSWORD = 'user1Password';

export async function runTest(projectDir: string, testModule: any) {
  await addApi(projectDir);
  updateSchemaInTestProject(projectDir, testModule.schema);
  await amplifyPush(projectDir);

  const awsconfig = configureAmplify(projectDir);
  const apiKey = getApiKey(projectDir);
  const appSyncClient = getConfiguredAppsyncClientAPIKeyAuth(awsconfig.aws_appsync_graphqlEndpoint, awsconfig.aws_appsync_region, apiKey);

  await testMutations(testModule, appSyncClient);
  await testQueries(testModule, appSyncClient);
}

export async function runAutTest(projectDir: string, testModule: any) {
  await addApi(projectDir, {
    'Amazon Cognito User Pool': {},
  });
  updateSchemaInTestProject(projectDir, testModule.schema);

  await updateAuthAddUserGroups(projectDir, [GROUPNAME]);
  await amplifyPush(projectDir);
  const awsconfig = configureAmplify(projectDir);

  const userPoolId = getUserPoolId(projectDir);
  await setupUser(userPoolId, USERNAME, PASSWORD, GROUPNAME);

  const user = await signInUser(USERNAME, PASSWORD);
  const appSyncClient = getConfiguredAppsyncClientCognitoAuth(awsconfig.aws_appsync_graphqlEndpoint, awsconfig.aws_appsync_region, user);

  await testMutations(testModule, appSyncClient);
  await testQueries(testModule, appSyncClient);
  await testSubscriptions(testModule, appSyncClient);
}

export async function runMultiAutTest(projectDir: string, testModule: any) {
  await addApi(projectDir, {
    'API key': {},
    'Amazon Cognito User Pool': {},
    IAM: {},
  });
  updateSchemaInTestProject(projectDir, testModule.schema);

  await updateAuthAddUserGroups(projectDir, [GROUPNAME]);
  await amplifyPush(projectDir);
  const awsconfig = configureAmplify(projectDir);

  const userPoolId = getUserPoolId(projectDir);
  await setupUser(userPoolId, USERNAME, PASSWORD, GROUPNAME);

  const user = await signInUser(USERNAME, PASSWORD);
  const appSyncClient = getConfiguredAppsyncClientCognitoAuth(awsconfig.aws_appsync_graphqlEndpoint, awsconfig.aws_appsync_region, user);

  await testMutations(testModule, appSyncClient);
  await testQueries(testModule, appSyncClient);
  await testSubscriptions(testModule, appSyncClient);
}

export function updateSchemaInTestProject(projectDir: string, schema: any) {
  const backendApiDirPath = path.join(projectDir, 'amplify', 'backend', 'api');
  const apiResDirName = fs.readdirSync(backendApiDirPath)[0];
  const backendSchemaFilePath = path.join(backendApiDirPath, apiResDirName, 'schema.graphql');
  fs.writeFileSync(backendSchemaFilePath, schema);
}

export async function testMutations(testModule: any, appSyncClient: any) {
  let mutationNames = Object.keys(testModule).filter(key => /^mutation[0-9]*$/.test(key));

  if (mutationNames.length > 1) {
    mutationNames = mutationNames.sort((name1, name2) => {
      const n1 = parseInt(name1.replace(/mutation/, ''));
      const n2 = parseInt(name2.replace(/mutation/, ''));
      return n1 - n2;
    });
  }

  const mutationTasks = [];
  mutationNames.forEach(mutationName => {
    const mutation = testModule[mutationName];
    const mutationInput = testModule['input_' + mutationName];
    const mutationResult = testModule['expected_result_' + mutationName];

    mutationTasks.push(async () => {
      await testMutation(appSyncClient, mutation, mutationInput, mutationResult);
    });
  });

  await runInSequential(mutationTasks);
}

export async function testMutation(appSyncClient: any, mutation: any, mutationInput?: any, mutationResult?: any) {
  let resultMatch = true;
  let errorMatch = true;
  let actualResponse;

  try {
    const result = await appSyncClient.mutate({
      mutation: gql(mutation),
      fetchPolicy: 'no-cache',
      variables: mutationInput,
    });
    if (!checkResult(result, mutationResult)) {
      actualResponse = result;
      resultMatch = false;
    }
  } catch (err) {
    if (!checkError(err, mutationResult)) {
      actualResponse = err;
      errorMatch = false;
    }
  }
  if (!resultMatch || !errorMatch) {
    console.log('The following mutation test failed.');
    console.log('Mutation: ', mutation);
    if (mutationInput) {
      console.log('Mutation input: ', mutationInput);
    }
    if (mutationResult) {
      console.log('Expected mutation result: ', mutationResult);
    }
    if (actualResponse) {
      console.log('Actual mutation response: ', actualResponse);
    }
    throw new Error('Mutation test failed.');
  }
}

export async function testQueries(testModule: any, appSyncClient: any) {
  let queryNames = Object.keys(testModule).filter(key => /^query[0-9]*$/.test(key));

  if (queryNames.length > 1) {
    queryNames = queryNames.sort((name1, name2) => {
      const n1 = parseInt(name1.replace(/query/, ''));
      const n2 = parseInt(name2.replace(/query/, ''));
      return n1 - n2;
    });
  }

  const queryTasks = [];
  queryNames.forEach(queryName => {
    const query = testModule[queryName];
    const queryInput = testModule['input_' + queryName];
    const queryResult = testModule['expected_result_' + queryName];

    queryTasks.push(async () => {
      await testQuery(appSyncClient, query, queryInput, queryResult);
    });
  });

  await runInSequential(queryTasks);
}

export async function testQuery(appSyncClient: any, query: any, queryInput?: any, queryResult?: any) {
  let resultMatch = true;
  let errorMatch = true;
  let actualResponse;
  try {
    const result = await appSyncClient.query({
      query: gql(query),
      fetchPolicy: 'no-cache',
      variables: queryInput,
    });
    if (!checkResult(result, queryResult)) {
      actualResponse = result;
      resultMatch = false;
    }
  } catch (err) {
    if (!checkError(err, queryResult)) {
      actualResponse = err;
      errorMatch = false;
    }
  }
  if (!resultMatch || !errorMatch) {
    console.log('The following query test failed.');
    console.log('Query: ', query);
    if (queryInput) {
      console.log('Query input: ', queryInput);
    }
    if (queryResult) {
      console.log('Expected query result: ', queryResult);
    }
    if (actualResponse) {
      console.log('Actual query response: ', actualResponse);
    }
    throw new Error('Query test failed.');
  }
}

export async function testSubscriptions(testModule: any, appsyncClient: any) {
  let subscriptionNames = Object.keys(testModule).filter(key => /^subscription[0-9]*$/.test(key));

  if (subscriptionNames.length > 1) {
    subscriptionNames = subscriptionNames.sort((name1, name2) => {
      const n1 = parseInt(name1.replace(/subscription/, ''));
      const n2 = parseInt(name2.replace(/subscription/, ''));
      return n1 - n2;
    });
  }

  const subscriptionTasks = [];
  subscriptionNames.forEach(subscriptionName => {
    const subscription = testModule[subscriptionName];
    const subscriptionInput = testModule['input_' + subscriptionName];
    const subscriptionResult = testModule['expected_result_' + subscriptionName];
    const mutations = testModule['mutations_' + subscriptionName];
    const mutationsInput = testModule['input_mutations_' + subscriptionName];

    subscriptionTasks.push(async () => {
      await testSubscription(appsyncClient, subscription, mutations, subscriptionResult, subscriptionInput, mutationsInput);
    });
  });

  await runInSequential(subscriptionTasks);
}

export async function testSubscription(
  appSyncClient: any,
  subscription: string,
  mutations: any[],
  subscriptionResult: any,
  subscriptionInput?: any,
  mutationInputs?: any[],
) {
  const observer = appSyncClient.subscribe({
    query: gql(subscription),
    variables: subscriptionInput,
  });

  const received = [];
  const sub = observer.subscribe((event: any) => {
    received.push(event.data);
  });

  await new Promise(res => setTimeout(() => res(), 4000));

  const mutationTasks = [];
  for (let i = 0; i < mutations.length; i++) {
    const mutation = mutations[i];
    const mutationInput = mutationInputs ? mutationInputs[i] : undefined;
    mutationTasks.push(async () => {
      await appSyncClient.mutate({
        mutation: gql(mutation),
        fetchPolicy: 'no-cache',
        variables: mutationInput,
      });
      await new Promise(res => setTimeout(() => res(), 4000)); //to ensure correct order in received data
    });
  }

  await runInSequential(mutationTasks);

  await new Promise(res => setTimeout(() => res(), 4000));

  sub.unsubscribe();
  if (!checkResult(received, subscriptionResult)) {
    console.log('The following subscription test failed.');
    console.log('Subscription: ', subscription);
    if (subscriptionResult) {
      console.log('Expected subscription result: ', subscriptionResult);
    }
    if (received) {
      console.log('Actual subscription response: ', received);
    }
    throw new Error('Subscription test failed.');
  }
}

/////
function checkResult(received: any, expected: any): boolean {
  if (!expected) {
    //the test does not request result check, as long as the mutation/query goes through, it's good
    return true;
  }
  const queue = [
    {
      received,
      expected,
      depth: 0,
    },
  ];
  try {
    return runCompare(queue);
  } catch (e) {
    console.log('checkResult error: ', e);
    return false;
  }
}

function checkError(received: any, expected: any): boolean {
  if (!expected) {
    //the test does not request result check, assume mutation/query should go through, but received error
    return false;
  }
  const queue = [
    {
      received,
      expected,
      depth: 0,
    },
  ];
  return runCompare(queue);
}

const MAX_DEPTH = 50;
function runCompare(queue: { received: any; expected: any; depth: number }[]): boolean {
  let result = true;

  while (queue.length > 0 && result) {
    const itemToCompare = queue.shift();
    if (itemToCompare.depth > MAX_DEPTH) {
      break;
    }
    if (typeof itemToCompare.expected === 'object') {
      if (itemToCompare.expected === null) {
        result = itemToCompare.received === null;
      } else if (itemToCompare.received === null) {
        result = false;
      } else if (typeof itemToCompare.received === 'object') {
        Object.keys(itemToCompare.expected).forEach(key => {
          queue.push({
            received: itemToCompare.received[key],
            expected: itemToCompare.expected[key],
            depth: itemToCompare.depth + 1,
          });
        });
      } else {
        result = false;
      }
    } else if (itemToCompare.expected === '<check-defined>') {
      result = itemToCompare.received !== null && itemToCompare.received !== undefined;
    } else {
      result = itemToCompare.received === itemToCompare.expected;
    }
  }

  return result;
}

export async function runInSequential(tasks: ((v: any) => Promise<any>)[]): Promise<any> {
  let result;

  for (const task of tasks) {
    result = await task(result);
  }

  return result;
}
