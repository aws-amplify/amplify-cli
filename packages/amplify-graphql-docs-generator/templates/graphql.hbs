# this is an auto generated file. This will be overwritten
{{#each operations }}
  {{> renderOp }}
{{/each}}
{{#each fragments }}
  {{> renderExternalFragment }}
{{/each}}