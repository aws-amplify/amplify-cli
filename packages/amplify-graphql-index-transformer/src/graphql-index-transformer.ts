import { DirectiveWrapper, InvalidDirectiveError, TransformerPluginBase } from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerResolverProvider,
  TransformerSchemaVisitStepContextProvider,
  TransformerTransformSchemaStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import {
  DirectiveNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InterfaceTypeDefinitionNode,
  Kind,
  ObjectTypeDefinitionNode,
} from 'graphql';
import { isScalarOrEnum } from 'graphql-transformer-common';
import { appendSecondaryIndex, constructSyncVTL, updateResolversForIndex } from './resolvers';
import { addKeyConditionInputs, ensureQueryField, updateMutationConditionInput } from './schema';
import { IndexDirectiveConfiguration } from './types';
import { validateNotSelfReferencing } from './utils';

const directiveName = 'index';
const directiveDefinition = `
  directive @${directiveName}(name: String!, sortKeyFields: [String], queryField: String) repeatable on FIELD_DEFINITION
`;

export class IndexTransformer extends TransformerPluginBase {
  private directiveList: IndexDirectiveConfiguration[] = [];
  private resolverMap: Map<TransformerResolverProvider, string> = new Map();

  constructor() {
    super('amplify-index-transformer', directiveDefinition);
  }

  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    context: TransformerSchemaVisitStepContextProvider,
  ): void => {
    const directiveWrapped = new DirectiveWrapper(directive);
    const args = directiveWrapped.getArguments({
      object: parent as ObjectTypeDefinitionNode,
      field: definition,
      directive,
    } as IndexDirectiveConfiguration);

    if (!args.sortKeyFields) {
      args.sortKeyFields = [];
    } else if (!Array.isArray(args.sortKeyFields)) {
      args.sortKeyFields = [args.sortKeyFields];
    }

    args.sortKey = [];

    validate(args, context as TransformerContextProvider);
    this.directiveList.push(args);
  };

  public after = (ctx: TransformerContextProvider): void => {
    if (!ctx.isProjectUsingDataStore()) return;

    // construct sync VTL code
    this.resolverMap.forEach((syncVTLContent, resource) => {
      if (syncVTLContent) {
        constructSyncVTL(syncVTLContent, resource);
      }
    });
  };

  transformSchema = (ctx: TransformerTransformSchemaStepContextProvider): void => {
    const context = ctx as TransformerContextProvider;

    for (const config of this.directiveList) {
      ensureQueryField(config, context);
      addKeyConditionInputs(config, context);
      updateMutationConditionInput(config, context);
    }
  };

  generateResolvers = (ctx: TransformerContextProvider): void => {
    for (const config of this.directiveList) {
      appendSecondaryIndex(config, ctx);
      updateResolversForIndex(config, ctx, this.resolverMap);
    }
  };
}

function validate(config: IndexDirectiveConfiguration, ctx: TransformerContextProvider): void {
  const { name, object, field, sortKeyFields } = config;

  validateNotSelfReferencing(config);

  const modelDirective = object.directives!.find(directive => {
    return directive.name.value === 'model';
  });

  if (!modelDirective) {
    throw new InvalidDirectiveError(`The @${directiveName} directive may only be added to object definitions annotated with @model.`);
  }

  config.modelDirective = modelDirective;

  const fieldMap = new Map<string, FieldDefinitionNode>();

  for (const objectField of object.fields!) {
    fieldMap.set(objectField.name.value, objectField);

    for (const peerDirective of objectField.directives!) {
      if (peerDirective === config.directive) {
        continue;
      }

      if (peerDirective.name.value === 'primaryKey') {
        config.primaryKeyField = objectField;

        if (
          objectField.name.value === field.name.value &&
          !peerDirective.arguments!.some((arg: any) => arg.name.value === 'sortKeyFields')
        ) {
          throw new InvalidDirectiveError(
            `Invalid @index '${name}'. You may not create an index where the partition key ` +
              'is the same as that of the primary key unless the primary key has a sort field. ' +
              'You cannot have a local secondary index without a sort key in the primary key.',
          );
        }
      }

      if (
        peerDirective.name.value === directiveName &&
        peerDirective.arguments!.some((arg: any) => arg.name.value === 'name' && arg.value.value === name)
      ) {
        throw new InvalidDirectiveError(
          `You may only supply one @${directiveName} with the name '${name}' on type '${object.name.value}'.`,
        );
      }
    }
  }

  const enums = ctx.output.getTypeDefinitionsOfKind(Kind.ENUM_TYPE_DEFINITION) as EnumTypeDefinitionNode[];

  if (!isScalarOrEnum(field.type, enums)) {
    throw new InvalidDirectiveError(`Index '${name}' on type '${object.name.value}.${field.name.value}' cannot be a non-scalar.`);
  }

  for (const sortKeyFieldName of sortKeyFields) {
    const sortField = fieldMap.get(sortKeyFieldName);

    if (!sortField) {
      throw new InvalidDirectiveError(
        `Can't find field '${sortKeyFieldName}' in ${object.name.value}, but it was specified in index '${name}'.`,
      );
    }

    if (!isScalarOrEnum(sortField.type, enums)) {
      throw new InvalidDirectiveError(
        `The sort key of index '${name}' on type '${object.name.value}.${sortField.name.value}' cannot be a non-scalar.`,
      );
    }

    config.sortKey.push(sortField);
  }
}
