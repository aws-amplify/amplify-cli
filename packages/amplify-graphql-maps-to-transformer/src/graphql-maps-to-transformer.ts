import { TransformerPluginBase, InvalidDirectiveError } from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerPluginType,
  TransformerSchemaVisitStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { ObjectTypeDefinitionNode, DirectiveNode, Kind, DefinitionNode } from 'graphql';
import { createMappingLambda } from './field-mapping-lambda';
import { attachFilterAndConditionInputMappingSlot, attachInputMappingSlot, attachResponseMappingSlot } from './field-mapping-resolvers';

const directiveName = 'mapsTo';

const directiveDefinition = `
  directive @${directiveName}(name: String!) on OBJECT
`;

export class MapsToTransformer extends TransformerPluginBase {
  constructor() {
    super(`amplify-maps-to-transformer`, directiveDefinition, TransformerPluginType.GENERIC);
  }

  /**
   * During the AST tree walking, the mapsTo transformer registers any renamed models with the ctx.resourceHelper
   */
  object = (definition: ObjectTypeDefinitionNode, directive: DirectiveNode, ctx: TransformerSchemaVisitStepContextProvider) => {
    const modelName = definition.name.value;
    const prevNameNode = directive.arguments?.find(arg => arg.name.value === 'name');

    const hasModelDirective = !!definition.directives?.find(directive => directive.name.value === 'model');
    if (!hasModelDirective) {
      throw new InvalidDirectiveError(`@mapsTo can only be used on an @model type`);
    }

    // the following checks should never fail because the graphql schema already validates them, but TS complains without them
    if (!prevNameNode) {
      throw new InvalidDirectiveError(`name is required in @${directiveName} directive`);
    }

    if (prevNameNode.value.kind !== 'StringValue') {
      throw new InvalidDirectiveError(`A single string must be provided for "name" in @${directiveName} directive`);
    }

    const originalName = prevNameNode.value.value;

    const schemaHasConflictingModel = !!ctx.inputDocument.definitions.find(hasModelWithNamePredicate(originalName));
    if (schemaHasConflictingModel) {
      throw new InvalidDirectiveError(`Type ${modelName} cannot map to ${originalName} because ${originalName} is a model in the schema.`);
    }
    ctx.resourceHelper.setModelNameMapping(modelName, originalName);
  };

  /**
   * During the generateResolvers step, the mapsTo transformer reads all of the model field mappings from the resourceHelper and generates
   * VTL to map the current field names to the original field names
   */
  after = (context: TransformerContextProvider) => {
    context.resourceHelper.getModelFieldMapKeys().forEach(modelName => {
      const modelFieldMap = context.resourceHelper.getModelFieldMap(modelName);
      if (!modelFieldMap.getMappedFields().length) {
        return;
      }
      const lambdaDataSource = createMappingLambda(context.api.host, context.stackManager);
      modelFieldMap.getResolverReferences().forEach(({ typeName, fieldName, isList }) => {
        const resolver = context.resolvers.getResolver(typeName, fieldName);
        if (!resolver) {
          return;
        }
        if (typeName === 'Mutation') {
          attachInputMappingSlot({
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
          });
          attachFilterAndConditionInputMappingSlot({
            slotName: 'preUpdate',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            dataSource: lambdaDataSource,
          });
          attachResponseMappingSlot({
            slotName: 'postUpdate',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            isList: false,
          });
        } else {
          attachFilterAndConditionInputMappingSlot({
            slotName: 'preDataLoad',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            dataSource: lambdaDataSource,
          });
          attachResponseMappingSlot({
            slotName: 'postDataLoad',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            isList,
          });
        }
      });
    });
  };
}

// returns a predicate for determining if a DefinitionNode is an model object with the given name
const hasModelWithNamePredicate = (name: string) => (node: DefinitionNode) =>
  node.kind === Kind.OBJECT_TYPE_DEFINITION && !!node.directives?.find(dir => dir.name.value === 'model') && node.name.value === name;
