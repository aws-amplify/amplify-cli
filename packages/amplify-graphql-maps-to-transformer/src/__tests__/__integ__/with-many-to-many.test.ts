import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { GraphQLTransform } from '@aws-amplify/graphql-transformer-core';
import { HasOneTransformer, ManyToManyTransformer } from '@aws-amplify/graphql-relational-transformer';
import { MapsToTransformer } from '../../graphql-maps-to-transformer';
import { IndexTransformer } from '@aws-amplify/graphql-index-transformer';
import { AuthTransformer } from '@aws-amplify/graphql-auth-transformer';
import { ObjectTypeDefinitionNode, parse } from 'graphql';
import { expectedResolversForModelWithRenamedField } from './common';

const manyToManyMapped = /* GraphQL */ `
  type Employee @model @mapsTo(name: "Person") {
    id: ID!
    tasks: [Task] @manyToMany(relationName: "EmployeeTask")
  }

  type Task @model @mapsTo(name: "Todo") {
    id: ID!
    title: String
    employees: [Employee] @manyToMany(relationName: "EmployeeTask")
  }
`;

const transformSchema = (schema: string) => {
  const indexTransformer = new IndexTransformer();
  const modelTransformer = new ModelTransformer();
  const hasOneTransformer = new HasOneTransformer();
  const authTransformer = new AuthTransformer();
  const transformer = new GraphQLTransform({
    transformers: [
      modelTransformer,
      indexTransformer,
      hasOneTransformer,
      authTransformer,
      new ManyToManyTransformer(modelTransformer, indexTransformer, hasOneTransformer, authTransformer),
      new MapsToTransformer(),
    ],
    sandboxModeEnabled: true,
  });
  return transformer.transform(schema);
};

describe('mapsTo with manyToMany', () => {
  it('creates resources with original GSIs and field names', () => {
    const out = transformSchema(manyToManyMapped);
    expect(out.stacks.EmployeeTask!.Resources!.EmployeeTaskTable.Properties.GlobalSecondaryIndexes).toMatchSnapshot();
    const outSchema = parse(out.schema);
    const EmployeeTaskFields = (
      outSchema.definitions.find(def => (def as any)?.name.value === 'EmployeeTask')! as ObjectTypeDefinitionNode
    ).fields!.map(field => field.name.value);
    expect(EmployeeTaskFields).toEqual(
      expect.arrayContaining(['id', 'employeeID', 'taskID', 'task', 'employee', 'createdAt', 'updatedAt']),
    );
    const expectedResolvers = expectedResolversForModelWithRenamedField('EmployeeTask');
    expectedResolvers.forEach(resolver => expect(out.resolvers[resolver]).toMatchSnapshot());
  });
});
