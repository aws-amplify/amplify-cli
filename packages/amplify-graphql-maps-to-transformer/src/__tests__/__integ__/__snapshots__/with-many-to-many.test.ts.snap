// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 1`] = `
Array [
  Object {
    "IndexName": "byPerson",
    "KeySchema": Array [
      Object {
        "AttributeName": "personID",
        "KeyType": "HASH",
      },
      Object {
        "AttributeName": "todoID",
        "KeyType": "RANGE",
      },
    ],
    "Projection": Object {
      "ProjectionType": "ALL",
    },
    "ProvisionedThroughput": Object {
      "Fn::If": Array [
        "ShouldUsePayPerRequestBilling",
        Object {
          "Ref": "AWS::NoValue",
        },
        Object {
          "ReadCapacityUnits": Object {
            "Ref": "DynamoDBModelTableReadIOPS",
          },
          "WriteCapacityUnits": Object {
            "Ref": "DynamoDBModelTableWriteIOPS",
          },
        },
      ],
    },
  },
  Object {
    "IndexName": "byTodo",
    "KeySchema": Array [
      Object {
        "AttributeName": "todoID",
        "KeyType": "HASH",
      },
      Object {
        "AttributeName": "personID",
        "KeyType": "RANGE",
      },
    ],
    "Projection": Object {
      "ProjectionType": "ALL",
    },
    "ProvisionedThroughput": Object {
      "Fn::If": Array [
        "ShouldUsePayPerRequestBilling",
        Object {
          "Ref": "AWS::NoValue",
        },
        Object {
          "ReadCapacityUnits": Object {
            "Ref": "DynamoDBModelTableReadIOPS",
          },
          "WriteCapacityUnits": Object {
            "Ref": "DynamoDBModelTableWriteIOPS",
          },
        },
      ],
    },
  },
]
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 2`] = `
"$util.qr($ctx.args.input.put(\\"personID\\", $ctx.args.input.employeeID))
$util.qr($ctx.args.input.remove(\\"employeeID\\"))
$util.qr($ctx.args.input.put(\\"todoID\\", $ctx.args.input.taskID))
$util.qr($ctx.args.input.remove(\\"taskID\\"))
$util.toJson({})"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 3`] = `
"#set( $fieldMap = {\\"employeeID\\":\\"personID\\",\\"taskID\\":\\"todoID\\"} )
#if( $util.isNull($fieldMap) || $fieldMap.keySet().size() <= 0 )
  #return($ctx.args)
#end
#if( $util.isNull($ctx.args.filter) && $util.isNull($ctx.args.condition) && $util.isNull($ctx.args.sort) && $util.isNull($ctx.args.aggregates) )
  #return($ctx.args)
#end
#set( $invoke = {
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"args\\": $ctx.args,
      \\"fieldMap\\": $fieldMap
  }
} )
$util.toJson($invoke)"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 4`] = `
"#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $ctx.stash.transformedArgs = $ctx.result )
$util.toJson({})"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 5`] = `
"$util.qr($ctx.prev.result.put(\\"employeeID\\", $ctx.prev.result.personID))
$util.qr($ctx.prev.result.remove(\\"personID\\"))
$util.qr($ctx.prev.result.put(\\"taskID\\", $ctx.prev.result.todoID))
$util.qr($ctx.prev.result.remove(\\"todoID\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 6`] = `
"$util.qr($ctx.args.input.put(\\"personID\\", $ctx.args.input.employeeID))
$util.qr($ctx.args.input.remove(\\"employeeID\\"))
$util.qr($ctx.args.input.put(\\"todoID\\", $ctx.args.input.taskID))
$util.qr($ctx.args.input.remove(\\"taskID\\"))
$util.toJson({})"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 7`] = `
"#set( $fieldMap = {\\"employeeID\\":\\"personID\\",\\"taskID\\":\\"todoID\\"} )
#if( $util.isNull($fieldMap) || $fieldMap.keySet().size() <= 0 )
  #return($ctx.args)
#end
#if( $util.isNull($ctx.args.filter) && $util.isNull($ctx.args.condition) && $util.isNull($ctx.args.sort) && $util.isNull($ctx.args.aggregates) )
  #return($ctx.args)
#end
#set( $invoke = {
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"args\\": $ctx.args,
      \\"fieldMap\\": $fieldMap
  }
} )
$util.toJson($invoke)"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 8`] = `
"#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $ctx.stash.transformedArgs = $ctx.result )
$util.toJson({})"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 9`] = `
"$util.qr($ctx.prev.result.put(\\"employeeID\\", $ctx.prev.result.personID))
$util.qr($ctx.prev.result.remove(\\"personID\\"))
$util.qr($ctx.prev.result.put(\\"taskID\\", $ctx.prev.result.todoID))
$util.qr($ctx.prev.result.remove(\\"todoID\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 10`] = `
"$util.qr($ctx.args.input.put(\\"personID\\", $ctx.args.input.employeeID))
$util.qr($ctx.args.input.remove(\\"employeeID\\"))
$util.qr($ctx.args.input.put(\\"todoID\\", $ctx.args.input.taskID))
$util.qr($ctx.args.input.remove(\\"taskID\\"))
$util.toJson({})"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 11`] = `
"#set( $fieldMap = {\\"employeeID\\":\\"personID\\",\\"taskID\\":\\"todoID\\"} )
#if( $util.isNull($fieldMap) || $fieldMap.keySet().size() <= 0 )
  #return($ctx.args)
#end
#if( $util.isNull($ctx.args.filter) && $util.isNull($ctx.args.condition) && $util.isNull($ctx.args.sort) && $util.isNull($ctx.args.aggregates) )
  #return($ctx.args)
#end
#set( $invoke = {
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"args\\": $ctx.args,
      \\"fieldMap\\": $fieldMap
  }
} )
$util.toJson($invoke)"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 12`] = `
"#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $ctx.stash.transformedArgs = $ctx.result )
$util.toJson({})"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 13`] = `
"$util.qr($ctx.prev.result.put(\\"employeeID\\", $ctx.prev.result.personID))
$util.qr($ctx.prev.result.remove(\\"personID\\"))
$util.qr($ctx.prev.result.put(\\"taskID\\", $ctx.prev.result.todoID))
$util.qr($ctx.prev.result.remove(\\"todoID\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 14`] = `
"$util.qr($ctx.prev.result.put(\\"employeeID\\", $ctx.prev.result.personID))
$util.qr($ctx.prev.result.remove(\\"personID\\"))
$util.qr($ctx.prev.result.put(\\"taskID\\", $ctx.prev.result.todoID))
$util.qr($ctx.prev.result.remove(\\"todoID\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 15`] = `
"#set( $fieldMap = {\\"employeeID\\":\\"personID\\",\\"taskID\\":\\"todoID\\"} )
#if( $util.isNull($fieldMap) || $fieldMap.keySet().size() <= 0 )
  #return($ctx.args)
#end
#if( $util.isNull($ctx.args.filter) && $util.isNull($ctx.args.condition) && $util.isNull($ctx.args.sort) && $util.isNull($ctx.args.aggregates) )
  #return($ctx.args)
#end
#set( $invoke = {
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"args\\": $ctx.args,
      \\"fieldMap\\": $fieldMap
  }
} )
$util.toJson($invoke)"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 16`] = `
"#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $ctx.stash.transformedArgs = $ctx.result )
$util.toJson({})"
`;

exports[`mapsTo with manyToMany creates resources with original GSIs and field names 17`] = `
"#foreach( $item in $ctx.prev.result.items )
  $util.qr($item.put(\\"employeeID\\", $item.personID))
  $util.qr($item.remove(\\"personID\\"))
  $util.qr($item.put(\\"taskID\\", $item.todoID))
  $util.qr($item.remove(\\"todoID\\"))
#end
$util.toJson($ctx.prev.result)"
`;
